### PR TITLE
DCS-1348 Adding date of birth validation

### PR DESCRIPTION
**Pull request by andrewrlee**

### CODE REVIEW

File: server/routes/bookedtoday/arrivals/searchforexisting/validation/dateOfBirthValidation.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  const day = parseInt(d, 10)
+  const month = parseInt(m, 10)
+  const year = parseInt(y, 10)
+
+  if (!day && !month && !year) {
+    return [{ text: "Enter this person's date of birth", href: '#date-of-birth-day' }]
+  }
+
+  if (day && month && year && !isValidDate(d, m, y)) {
+    return [{ text: "Enter this person's date of birth", href: '#date-of-birth-day' }]
```
Comment (andrewrlee) [inline server/routes/bookedtoday/arrivals/searchforexisting/validation/dateOfBirthValidation.ts:22]: awaiting text for this

File: server/routes/bookedtoday/arrivals/searchforexisting/validation/dateOfBirthValidation.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+  const missingFieldNames = zip(fields, [day, month, year])
+    .map(([field, value]) => !value && field)
+    .filter(Boolean)
+
+  const message = missingFieldNames.join(' and ')
+
+  return !missingFieldNames.length
+    ? []
+    : [{ text: `Date of birth must include a ${message}`, href: `#date-of-birth-${missingFieldNames[0]}` }]
```
Comment (andrewrlee) [inline server/routes/bookedtoday/arrivals/searchforexisting/validation/dateOfBirthValidation.ts:33]: point to first invalid field

File: server/routes/bookedtoday/arrivals/searchforexisting/validation/nameValidation.test.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+      },
+    ]))
+
+  it('Last name required', () =>
+    expect(NameValidator({ firstName: 'Sam' })).toEqual([
+      {
+        href: '#last-name',
+        text: "Enter this person's last name",
+      },
+    ]))
```
Comment (andrewrlee) [inline server/routes/bookedtoday/arrivals/searchforexisting/validation/nameValidation.test.ts:18]: adding missing test

File: server/views/pages/bookedtoday/arrivals/searchForExistingRecord/changeDateOfBirth.njk [hunk 4]
```diff
@@ -15,21 +22,22 @@
                         hint: {
                             text: "For example, 31 3 1980."
                         },
+                        errorMessage: errors | findFirstError(["date-of-birth-day", "date-of-birth-month", "date-of-birth-year"]),
                         items: [
                             {
                             classes: "govuk-input--width-2",
                             name: "day",
-                            value: data.dateOfBirth | formatDate('DD')
+                            value: data.day | default(data.dateOfBirth | formatDate('DD'))
```
Comment (andrewrlee) [inline server/views/pages/bookedtoday/arrivals/searchForExistingRecord/changeDateOfBirth.njk:30]: read from input from flash if present else default to data from cookie

File: server/routes/bookedtoday/arrivals/searchforexisting/changeDateOfBirthController.ts [hunk 5]
```diff
@@ -4,19 +4,30 @@ import { State } from './state'
...
+        return res.redirect(`/prisoners/${id}/search-for-existing-record/change-date-of-birth`)
+      }
+
       const { day, month, year } = req.body
-      // TODO build and validate
-      State.searchDetails.update(req, res, { dateOfBirth: `${year}-${month}-${day}` })
 
-      res.redirect(`/prisoners/${id}/search-for-existing-record`)
+      State.searchDetails.update(req, res, {
+        dateOfBirth: `${year.padStart(4, '0')}-${month.padStart(2, '0')}-${day.padStart(2, '0')}`,
```
Comment (andrewrlee) [inline server/routes/bookedtoday/arrivals/searchforexisting/changeDateOfBirthController.ts:27]: safe to build this now as validated - need to pad year or moment starts getting fancy and interpreting the fields in the wrong way - even with strict mode

**Review by louisenorris — APPROVED** [review]
LGTM

**Pull request merged**